### DISCUSSION TITLE
Saving and reloading an EB geometry from chk_geom

### DISCUSSION THREAD

**Discussion by jagmohas** [AMReX-Combustion/PeleC#638] [Q&A] [answered]
I am trying to use Saving and Reloading an EB_geometry from a check point file. I am trying this with EB-C10 example. Following the instructions from 
https://amrex-combustion.github.io/PeleC/geometry/EB.html

I added eb2.write_chk_geom=1 in the example.inp. this writes the chk_geom file. 
In order to read this chk_geom file, I added 
eb2.geom_type="chkfile"
eb2.chkfile="chk_geom" 

It fails with this error
AMReX/Src/EB/AMReX_EB_chkpt_file.cpp", line 191, Msg: "EB2::ChkptFile cannot read from different max_grid_size" !!!

I have not changed anything else in the example.inp. May I ask what I might be doing wrong?

Thank you!

**Comment by marchdf** [comment]
I see the same thing... let me see if I can figure out why.

**Reply by marchdf** [reply, marked as answer]
Can you add this to the input file: `eb2.max_grid_size=32` ? It fixed it on my end. If that works, I will update the docs in PR #639

**Reply by jagmohas** [reply]
That works. Thank you! 
I tried with 16, 32, 64 and note that we need to give eb2.max_grid_size same as amr.max_grid_size then only it works.

**Reply by marchdf** [reply]
yeah it's supposed to match the one used when generating the EB in the first place. It defaults to 64, and that's what we've been using it with fortuitously.

**Reply by jagmohas** [reply]
Thank you, Marc. 
If I understand correctly, this feature is only effective when we maintain the same AMR (Adaptive Mesh Refinement) levels as in the 'chk_geom' file, is that correct? In our simulations, we might start with a lower 'max_level' and gradually increase the AMR levels. Does this mean that every time we raise the AMR levels, we should regenerate the 'chk_geom' file? I've also observed that regardless of the number of AMR levels used, the 'chk_geom' file contains data only under 'Level_0' (without any 'Level_1' or 'Level_2' directories). I assume that all the necessary geometric data is written under 'Level_0', otherwise my tests with 'max_level' greater than 0 would fail and I would be getting incorrect geometry.

**Reply by drummerdoc** [reply]
It's a bit of a complicated answer, and that's because in AMReX the geometry generation is intimately tied to the geometric multigrid linear solvers (coarsened versions of the geometry are used in building the mg v-cycles) that you probably aren't using with PeleC, and for that there are benefits to having the volumes and edge properly telescope across these coarsened levels.  So, there are restrictions you are hitting with that may make little sense to you as a PeleC user.  However, the concepts apply even in PeleC when finer levels are added...we want the volumes and areas to properly telescope with levels (i.e., the fine cells at level 3 contributing to a cell at level 2 should have volumes that sum to the level 2 cell volume etc).  This means that level 2 EB cells can be different whether they are generated geometrically at level 2 or generated by coarsening a geometry generated geometrically at level 3.  There is a flag to control whether that type of coarsening indeed happens, but it's an option you usually don't want to employ...conservation is usually pretty critical in these AMR calcs.

One way to consistently manage what you want to do is to generate the geometry to the finest max_level you intend for that problem, but set up the refinement criteria initially to simply never generate AMR data at that level until you are ready to do so.  You will be carrying around a much larger EB geometry than needed for a while, but when you add levels, they will be consistent.  And with the EB IO capability, you can decide whether it is more efficient to generate once and read/write all the time, or regen on every restart.

Another way to simplify your life is to shut off the generation of coarsened EB levels (coarser than level 0) that you won't be using.
